llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp

```diff
@@ -1752,6 +1752,15 @@ bool SIInsertWaitcnts::generateWaitcntInstBefore(MachineInstr &MI,
         const bool IsVGPR = TRI->isVectorRegister(*MRI, Op.getReg());
         for (int RegNo = Interval.first; RegNo < Interval.second; ++RegNo) {
           if (IsVGPR) {
+            // Implicit VGPR defs and uses are never a part of the memory
+            // instructions description and usually present to account for
+            // super-register liveness. Tied implicit sources on loads though
+            // are real uses.
+            // TODO: Most of the other instructions also have implicit uses
+            // for the liveness accounting only.
+            if (Op.isImplicit() && MI.mayLoadOrStore() && !Op.isTied())
```

Comment:
I'm a bit nervous because lots of instructions are marked mayLoad/mayStore just because they have side effects, not because they are regular load or store instructions.

I _think_ you can ignore tied uses here because we will still handle the WAW conflict (if any) based on the real def operand that it is tied to.

Reply:
I don't see why mayLoadOrStore matters, implicit operands on any instruction shouldn't matter

Reply:
> I'm a bit nervous because lots of instructions are marked mayLoad/mayStore just because they have side effects, not because they are regular load or store instructions.

Do you want a check like isVMEM() || isFLAT() instead?

> I _think_ you can ignore tied uses here because we will still handle the WAW conflict (if any) based on the real def operand that it is tied to.

I've a regression this way in one of the tests with buffer_load and tied def. The wait was lost, hence I've added the check.

Reply:
> I don't see why mayLoadOrStore matters, implicit operands on any instruction shouldn't matter

Implicit operands matter at least with movrel. The check for mayLoadOrStore is really to catch spills which are a real problem. Potentially a very wide loads and stores a problem too.

Reply:
> I've a regression this way in one of the tests with buffer_load and tied def. The wait was lost, hence I've added the check.

What was the test? Are you sure the wait was required? For a WAW-like dependency, no wait is required if the loads are guaranteed to complete (and write their vgpr results) in order.

Reply:
The failed test was image-waterfall-loop-O0.ll, this wait was missing:
```
   v_mov_b32_e32 v3, s4
   ; kill: killed $vgpr4
   s_xor_saveexec_b32 s4, -1
   s_waitcnt vmcnt(0)
   buffer_load_dword v0, off, s[0:3], s32 offset:80 ; 4-byte Folded Reload
   buffer_load_dword v2, off, s[0:3], s32 offset:84 ; 4-byte Folded Reload
```
That is how pass debug log looks if I remove the tied check:
```
    VM_CNT(2): 1:v0 0:v4
    LGKM_CNT(0):
    EXP_CNT(0):
    VS_CNT(86):

$vgpr0 = BUFFER_LOAD_DWORD_OFFSET $sgpr0_sgpr1_sgpr2_sgpr3, $sgpr32, 80, 0, 0, implicit $exec, implicit $vgpr0(tied-def 0) :: (load (s32) from %stack.16, addrspace 5)
```
So it reads v0 and merges the load back. This may be not needed for a dword load, but what if we read 16-bit and preserve the other half? The pattern will be the same, a tied def.

Reply:
It may be possible to write something like this though:
```
            if (Op.isImplicit() && MI.mayLoadOrStore() && (!Op.isTied() || ST->hasVmemWriteVgprInOrder()))
              continue;
```
but this looks too scary.

Reply:
> So it reads v0 and merges the load back. This may be not needed for a dword load, but what if we read 16-bit and preserve the other half? The pattern will be the same, a tied def.

All these cases are handled elsewhere as WAW hazards if necessary (usually it is not necessary because buffer loads complete in order anyway). You do not need to handle them here. The code here is only for RAW hazards.

Reply:
> > So it reads v0 and merges the load back. This may be not needed for a dword load, but what if we read 16-bit and preserve the other half? The pattern will be the same, a tied def.
> 
> All these cases are handled elsewhere as WAW hazards if necessary (usually it is not necessary because buffer loads complete in order anyway). You do not need to handle them here. The code here is only for RAW hazards.

Why don't we have FeatureVmemWriteVgprInOrder in gfx12?

Reply:
Because in GFX12, vmem loads do _not_ write their vgpr results in order. See #105549. (When I said "usually it is not necessary" I meant pre-GFX12.)

Reply:
> Because in GFX12, vmem loads do _not_ write their vgpr results in order. See #105549. (When I said "usually it is not necessary" I meant pre-GFX12.)

So you mean is there is SCRATCH_LOAD_USHORT merging vdst on gfx12 it will be handled anyway?

Reply:
Just did it, removed the check for tied operand.

Reply:
SCRATCH_LOAD_USHORT writes all 32 bits, but if you mean a case like this:
```
scratch_load_dword v0, ...
scratch_load_short_d16 v0, ... // merge into low 16 bits of v0
```
... then yes I think we will insert a wait on GFX12 to avoid the WAW, and this is done only by looking at the defs, it does not depend on any tied uses. And pre-GFX12 I don't think any wait is required for this case.

Reply:
I will try to write a mir test for this. Say, scratch_load_dword followed by a scratch_load_short_d16 into the same register. My understanding is: it shall not need wait before gfx12, but needs it on gfx12.

Reply:
Added the test. It turns out d16 loads are properly handled and the tied operand is not implicit.